src/drstore.jl

```diff
@@ -0,0 +1,74 @@
+###
+# special struct to hold decision rules
+###
+
+# import Main
+struct DRStore{T}
+    data::Tuple{Vararg{Vector{T}}}
+    flat::Vector{T}
+end
+
+Main.length(ds::DRStore) = length(ds.data)
+Main.size(ds::DRStore, x...) = size(ds.data, x...)
+Main.getindex(ds::DRStore,x...) = getindex(ds.data,x...)
+Main.setindex!(ds::DRStore,x...) = setindex!(ds.data,x...)
+Main.abs(ds::DRStore{T}) where T = abs(ds.flat)
+
+Main.maxabs(ds::DRStore{T}) where T = maximum(e->norm(e,Inf), ds.flat)
+
+Main.norm(ds::DRStore{T}) where T = maximum(e->norm(e,Inf), ds.flat)
+# Main.norm(ds::DRStore{T},k::Int) where T = norm( (norm(e,k) for e in ds.flat), k)
+distance(ds1::DRStore{T}, ds2::DRStore{T}) where T = maximum( k->norm(k[1]-k[2],Inf), zip(ds1.flat,ds2.flat) )
+# distance(ds1::DRStore{T}, ds2::DRStore{T}, k::Int) where T = norm( (norm(e[1]-e[2],k) for e in zip(ds1.flat,ds2.flat)), k )
+
+total(ds::DRStore{T}) where T = sum(length(e) for e in ds.data)
+
+function DRStore(::Type{T}, dims::AbstractVector{Int}) where T
+    L = sum(dims)
+    fD = zeros(T, L)
+    ptr = pointer(fD)
```

Comment:
Yikes 😨 ! I don't think I'v ever seen `pointer` and `unsafe_wrap` used out in in the pure Julia wild (e.g. not interfacing with C libs) 

Are we confident that this is safe and works correctly? I would guess that it does given your experiments, but it makes me a little scared.

What type of penalty would we get if we were to construct `D` using SubArrays (e.g. using `view`) instead of `unsafe_wrap` and pointer offsets?

Reply:
I think problem is not whether it is safe:
- we control object construction, and the object is immutable so user cannot alter its behaviour. Of course this assumes we add consistency checks.
- all arrays are linked to the same object, so one "view" cannot be destroyed without the other being destroyed.  By construction, the memory we access should be freed for all arrays/subarrays at the same time.
One potential problem is actually performance. From my reptilian understanding of the discussion here (https://github.com/JuliaLang/julia/issues/22849), the compiler can optimize further if we guarantee non-aliasing. Se we might slow down things a little bit with this approach (but not more than our current use of reinterpret).
In principle, SubArray should work, and we should be able to pass them to all routines (including interpolation routines). Right now it doesn't:
- our interpolation routine assume that input are contiguous arrays (i.e. take Array as arguments). We rely on this to reinterpret arrays in both directions.
- we cannot just change all arguments to AbstractArray. That would be inapropriate, because it doesn't give us any guarantee that we can reinterpret them. 
- SubArray have a flag which tells whether they are contiguous, but reinterpret doesn't work on them. 
Intuitively, the condition for us to be able to reinterpret an array, should not for it to be contiguous but for all consecutive chunks to be a multiple of the size of all datatype we wish to reinterpret it into.  I think this is the direction, which is followed in Julia 0.7, where reinterpret will return a special type of Array.  It would be natural that some SubArrays are reinterpretable too, but I don't know whether this is the case yet.

Right now, I don't totally understand how the new reintepret will work in 0.7 and it certainly won't be backward compatible.
On the other hand, the current behaviour is arguably stupid, but certainly forward compatible, unless they remove pointer, which I'm sure they won't. Maybe there is a way to actually tell the compiler we are not aliasing anything. Overall there are less factors for us to take into account with that simple solution so I'm quite inclined to stay with it for now.